app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt

```diff
@@ -301,30 +310,88 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
     }
   }
 
-  private fun onBookItemClick(item: LibraryListItem.BookItem) {
-    when {
-      isNotConnected -> {
-        noInternetSnackbar()
-        return
+  private fun checkExternalStorageWritePermission(): Boolean {
+    return hasPermission(WRITE_EXTERNAL_STORAGE).also { permissionGranted ->
+      if (!permissionGranted) {
+        if (shouldShowRationale(WRITE_EXTERNAL_STORAGE)) {
+          alertDialogShower.show(
+            KiwixDialog.WriteStoragePermissionRationale,
+            ::requestExternalStoragePermission
+          )
+        } else {
+          alertDialogShower.show(
+            KiwixDialog.WriteStoragePermissionRationale,
+            ::openAppSettings
+          )
+        }
       }
-      noWifiWithWifiOnlyPreferenceSet -> {
-        dialogShower.show(WifiOnly, {
-          sharedPreferenceUtil.putPrefWifiOnly(false)
-          downloadFile(item.book)
-        })
-        return
+    }
+  }
+
+  private fun openAppSettings() {
+    val uri: Uri = Uri.fromParts("package", requireActivity().packageName, null)
+    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
+      flags = Intent.FLAG_ACTIVITY_NEW_TASK
+      data = uri
+    }
+    startActivity(intent)
+  }
+
+  private fun requestExternalStoragePermission() {
+    ActivityCompat.requestPermissions(
+      requireActivity(), arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+      REQUEST_STORAGE_PERMISSION
+    )
+  }
+
+  private fun shouldShowRationale(writeExternalStorage: String) =
+    ActivityCompat.shouldShowRequestPermissionRationale(requireActivity(), writeExternalStorage)
+
+  override fun onRequestPermissionsResult(
+    requestCode: Int,
+    permissions: Array<out String>,
+    grantResults: IntArray
+  ) {
+    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+    if (requestCode == REQUEST_STORAGE_PERMISSION &&
+      grantResults.isNotEmpty() &&
+      permissions[0] == Manifest.permission.WRITE_EXTERNAL_STORAGE
```

Comment:
In here `permissions` may be null. Instead of checking `grantResults.isNotEmpty()`, we should check `permissions.isNotEmpty()`. We don't really need `grantResults` here.

Reply:
done